pkg/internal/metrics/metrics.go

```diff
@@ -0,0 +1,198 @@
+	//
+	// This is a gauge rather than a counter because certificate requests may
+	// get removed over time e.g. with revisionHistoryLimit.
+	approvedCount = prometheus.NewDesc(
+		"approverpolicy_certificaterequest_approved_count",
+		"Number of CertificateRequests that were been approved or denied (Approved=True or Approved=False).",
+		[]string{
+			"namespace",
+			"approved_status",
+		},
+		nil,
+	)
+
+	// orphanCount counts the current number of certificate requests that have
+	// not been matched by any approvers. We call then "orphans". Orphans don't
+	// have the Approved condition.
+	orphanCount = prometheus.NewDesc(
+		"approverpolicy_certificaterequest_orphan_count",
+		"Number of orphan CertificateRequests, i.e., that don't have an Approved condition set.",
+		[]string{
+			"name",
+			"namespace",
+		},
+		nil,
+	)
+)
+
+// You don't need to wait for the cache to be synced before calling this. This
+// function is non-blocking.
+func RegisterMetrics(ctx context.Context, log logr.Logger, c cache.Cache) {
+	metrics.Registry.MustRegister(collector{ctx, log, c})
+}
+
+// We use a custom collector instead of prometheus.NewGaugeVec because it is
+// much easier to list all of the certificate requests when the `/metrics`
+// endpoint is hit rather than using a controller-runtime reconciler.
+type collector struct {
+	ctx   context.Context
+	log   logr.Logger
+	cache cache.Cache
+}
+
+func (cc collector) Describe(ch chan<- *prometheus.Desc) {
+	prometheus.DescribeByCollect(cc, ch)
+}
+
+func (cc collector) Collect(ch chan<- prometheus.Metric) {
+	// We found a niche problem where `/metrics` would hang forever in case of a
+	// misconfigured RBAC. This was due to `cache.List` hanging until the cache
+	// is synced. To prevent that, we skip reporting this subset of the metrics.
+	if !hasSynced(cc.cache) {
+		cc.log.Info("cache not synced yet, skipping metrics approverpolicy_certificaterequest_*")
+		return
+	}
+
+	collectCRsApproved(cc.ctx, cc.log, cc.cache, ch)
+	collectCRsOrphans(cc.log, cc.cache, ch)
+}
+
+// hasSynced returns true if the cache has synced. Otherwise, it returns false.
+// It is non-blocking.
+func hasSynced(cache cache.Cache) bool {
+	// Why tempCtx? Controller-runtime doesn't give us a non-blocking way to
+	// know if the cache is in sync. The only way it to use `WaitForCacheSync`,
+	// but that's a blocking operation. 10 milliseconds seems like plenty enough
+	// time since `WaitForCacheSync` doesn't do any network call to know if the
+	// cache is ready.
+	tempCtx, cancel := context.WithTimeout(context.Background(), 10*time.Millisecond)
+	defer cancel()
+	return cache.WaitForCacheSync(tempCtx)
+}
+
+func collectCRsApproved(ctx context.Context, log logr.Logger, c cache.Cache, ch chan<- prometheus.Metric) {
+	list := &cmapi.CertificateRequestList{}
+	err := c.List(ctx, list)
+	if err != nil {
+		log.Error(err, "unable to list CertificateRequests")
+		return
+	}
+
+	type label struct {
+		namespace, approvedStatus string
```

Comment:
```suggestion
		namespace string
		approvedStatus metav1.ConditionStatus
```

I would prefer if we keep the `ConditionStatus` type alias until the metrics are produced at the end of this function. IMO it would make the code more readable.

Reply:
Done!